import asyncio
import os
import signal
import traceback
from datetime import datetime

import croniter
import uvloop
from aiorun import run
from pyrogram.errors import (AuthKeyDuplicated, AuthKeyUnregistered,
                             SessionRevoked, UserAlreadyParticipant,
                             UserDeactivated, UserDeactivatedBan)
from pytz import timezone

from config import LOG_SELLER, OWNER_ID
from Haji import UserBot, bot, logger, haji
from Haji.database import dB
from Haji.helpers import (CheckUsers, CleanAcces, ExpiredUser, Tools,
                          check_payment, installPeer)

list_error = []


asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())


async def cleanup_total(ubot_id):
    """Clean up database records for a specific userbot."""
    try:
        await dB.rem_expired_date(ubot_id)
        await dB.rem_pref(ubot_id)
        await dB.rm_all(ubot_id)
        await dB.remove_ubot(ubot_id)
        logger.info(f"Deleted user {ubot_id}")
    except Exception as e:
        logger.error(f"Failed to cleanup userbot {ubot_id}: {e}")


async def handle_start_error():
    if list_error:
        for data in list_error:
            ubot = data["user"]
            reason = data["error_msg"]
            await bot.send_message(
                LOG_SELLER,
                f"<b>Userbot {ubot} failed to start due to {reason}, deleted user on database</b>",
            )
            await cleanup_total(ubot)


async def start_ubot(ubot):
    """Start a userbot instance and handle setup."""
    userbot = UserBot(**ubot)
    try:
        await userbot.start()
        sudo_users = await dB.get_list_from_var(userbot.me.id, "SUDOERS")
        for user in sudo_users:
            userbot.add_sudoers(userbot.me.id, user)
            userbot.add_sudoers(userbot.me.id, userbot.me.id)
        for chat in ["MediaUmum", "GcSupportMek", "OfficialFreesex"]:
            try:
                await userbot.join_chat(chat)
            except UserAlreadyParticipant:
                pass
            except Exception:
                continue
    except (AuthKeyUnregistered, AuthKeyDuplicated, SessionRevoked):
        reason = "Session Ended"
        data = {"user": int(ubot["name"]), "error_msg": reason}
        list_error.append(data)
    except (UserDeactivated, UserDeactivatedBan):
        reason = "Account Banned by Telegram"
        data = {"user": int(ubot["name"]), "error_msg": reason}
        list_error.append(data)


async def start_main_bot():
    """Start the main bot after userbots."""
    logger.info("ü§ñ Starting main bot...")
    await bot.start()
    await bot.add_reseller()
    total_bots = len(haji._ubot)
    message = "üî•**Userbot berhasil diaktifkan**üî•\n" f"‚úÖ **Total User: {total_bots}**"
    await dB.set_var(bot.id, "total_users", total_bots)
    logger.info("‚úÖ Main bot started successfully.")
    try:
        await bot.send_message(OWNER_ID, f"<blockquote>{message}</blockquote>")
    except Exception:
        pass


async def start_userbots():
    await dB.initialize()
    logger.info("üîÑ Starting userbots...")
    userbots = await dB.get_userbots()
    tasks = [asyncio.create_task(start_ubot(ubot)) for ubot in userbots]
    results = await asyncio.gather(*tasks, return_exceptions=False)

    for idx, result in enumerate(results):
        if isinstance(result, Exception):
            logger.error(f"‚ùå Error starting userbot {userbots[idx]['name']}: {result}")
            await bot.send_message(
                LOG_SELLER,
                f"‚ùå Error starting userbot {userbots[idx]['name']}: {result}",
            )

    logger.info("‚úÖ All userbots started successfully.")


async def end_main():
    background_tasks = [
        ExpiredUser(),
        installPeer(),
        CheckUsers(),
        CleanAcces(),
    ]
    for task in background_tasks:
        asyncio.create_task(task)
    logger.info(f"Started {len(background_tasks)} background tasks")


async def stop_main():
    tasks = [t for t in asyncio.all_tasks() if t is not asyncio.current_task()]
    logger.info(f"üìå Total task yang akan dihentikan: {len(tasks)}")

    if not tasks:
        logger.info("‚úÖ Tidak ada task yang berjalan.")
        return

    for task in tasks:
        if not task.done():
            task.cancel()

    await asyncio.sleep(1)

    results = []
    for task in tasks:
        try:
            result = await asyncio.wait_for(asyncio.shield(task), timeout=5)
            results.append(result)
        except asyncio.TimeoutError:
            logger.error(
                f"‚è≥ Timeout saat menghentikan task: {task.get_name() or task}"
            )
            results.append(None)
        except asyncio.CancelledError:
            continue
        except Exception as e:
            logger.error(f"‚ö†Ô∏è Task {task.get_name() or task} mengalami error: {e}")

    logger.info("‚ö†Ô∏è Menutup database...")
    await Tools.close_fetch()
    await dB.close()
    logger.info("‚úÖ Semua task dihentikan dan database telah ditutup.")


async def main():
    try:
        await start_userbots()
        await start_main_bot()
        await end_main()
        await handle_start_error()

    except asyncio.CancelledError:
        logger.warning("Stopped All.")
    except Exception:
        logger.error(f"{traceback.format_exc()}")


if __name__ == "__main__":
    run(
        main(),
        loop=bot.loop,
        shutdown_callback=stop_main(),
    )
